types/src/v2/params.rs

```diff
@@ -68,7 +68,90 @@ impl<'a> RpcParams<'a> {
 		Self(raw)
 	}
 
-	/// Attempt to parse all parameters as array or map into type T
+	fn next_inner<T>(&mut self) -> Option<Result<T, CallError>>
+	where
+		T: Deserialize<'a>,
+	{
+		let mut json = self.0?.trim_start();
+
+		match json.as_bytes().get(0)? {
+			b']' => {
+				self.0 = None;
+
+				return None;
+			}
+			b'[' | b',' => json = &json[1..],
+			_ => return Some(Err(CallError::InvalidParams)),
+		}
+
+		let mut iter = serde_json::Deserializer::from_str(json).into_iter::<T>();
+
+		match iter.next()? {
+			Ok(value) => {
+				self.0 = Some(&json[iter.byte_offset()..]);
+
+				Some(Ok(value))
+			}
+			Err(_) => {
+				self.0 = None;
+
+				Some(Err(CallError::InvalidParams))
+			}
+		}
+	}
+
+	/// Parse the next parameter to type `T`
+	///
+	/// ```
+	/// # use jsonrpsee_types::v2::params::RpcParams;
+	/// let mut params = RpcParams::new(Some(r#"[true, 10, "foo"]"#));
+	///
+	/// let a: bool = params.next().unwrap();
+	/// let b: i32 = params.next().unwrap();
+	/// let c: &str = params.next().unwrap();
+	///
+	/// assert_eq!(a, true);
+	/// assert_eq!(b, 10);
+	/// assert_eq!(c, "foo");
+	/// ```
+	pub fn next<T>(&mut self) -> Result<T, CallError>
+	where
+		T: Deserialize<'a>,
+	{
+		match self.next_inner() {
+			Some(result) => result,
+			None => Err(CallError::InvalidParams),
+		}
+	}
+
+	/// Parse the next optional parameter to type `Option<T>`.
+	///
+	/// The result will be `None` for `null`, and for missing values in the supplied JSON array.
+	///
+	/// ```
+	/// # use jsonrpsee_types::v2::params::RpcParams;
+	/// let mut params = RpcParams::new(Some(r#"[1, 2, null]"#));
+	///
+	/// let params: [Option<u32>; 4] = [
+	///     params.optional_next().unwrap(),
+	///     params.optional_next().unwrap(),
+	///     params.optional_next().unwrap(),
+	///     params.optional_next().unwrap(),
+	/// ];;
+	///
+	/// assert_eq!(params, [Some(1), Some(2), None, None]);
```

Comment:
Does this work when calling `parse()` directly? Like

```rust
let in_params = RpcParams::new(Some(r#"[1, 2, null]"#));
let params: [Option<u32>; 4] = in_params.parse().unwrap();
assert_eq!(params, [Some(1), Some(2), None, None]);
```
?

Reply:
We'd need trait specialization to be able to do that unfortunately.